packages/lit-analyzer/src/analyze/util/type/is-assignable-binding-under-security-system.ts

```diff
@@ -0,0 +1,139 @@
+interface SecurityOverrideMap {
+	[attributeName: string]: string[] | undefined;
+}
+
+const closureScopedOverrides: TagNameToSecurityOverrideMap = {
+	iframe: {
+		src: ["TrustedResourceUrl"]
+	},
+	a: {
+		href: ["TrustedResourceUrl", "SafeUrl", "string"]
+	},
+	img: {
+		src: ["TrustedResourceUrl", "SafeUrl", "string"]
+	},
+	script: {
+		src: ["TrustedResourceUrl"]
+	}
+};
+const closureGlobalOverrides: SecurityOverrideMap = {
+	style: ["SafeStyle", "string"]
+};
+
+function checkClosureSecurityAssignability(
+	typeB: SimpleType,
+	htmlAttr: HtmlNodeAttr,
+	request: LitAnalyzerRequest,
+	source: LitAnalyzerRuleName
+): LitHtmlDiagnostic[] | undefined {
+	const scopedOverride = closureScopedOverrides[htmlAttr.htmlNode.tagName];
+	const overriddenTypes = (scopedOverride && scopedOverride[htmlAttr.name]) || closureGlobalOverrides[htmlAttr.name];
+	if (overriddenTypes === undefined) {
+		return undefined;
+	}
+	const typeMatch = matchesAtLeastOneNominalType(overriddenTypes, typeB);
+	if (typeMatch === undefined) {
+		return undefined;
+	}
+	if (typeMatch === false) {
+		const nominalType: SimpleType = {
+			kind: SimpleTypeKind.INTERFACE,
+			members: [],
+			name: "A security type"
+		};
+		return [
+			{
+				kind: LitHtmlDiagnosticKind.INVALID_ATTRIBUTE_EXPRESSION_TYPE,
+				message: `Type '${toTypeString(typeB)}' is not assignable to '${overriddenTypes.join(" | ")}'. This is due to Closure Safe Type enforcement.`,
+				severity: "error",
+				source,
+				location: { document: request.document, ...htmlAttr.location.name },
+				htmlAttr,
+				typeA: nominalType,
+				typeB
+			}
+		];
+	}
+	return [];
+}
+
+function matchesAtLeastOneNominalType(typeNames: string[], typeB: SimpleType): boolean | undefined {
+	// Directives are out of scope of lit's security system.
+	if (looksLikeADirective(typeB)) {
+		return undefined;
+	}
+	if (typeB.name !== undefined && typeNames.includes(typeB.name)) {
+		return true;
+	}
+	switch (typeB.kind) {
+		case "UNION":
+			return typeB.types.every(t => matchesAtLeastOneNominalType(typeNames, t));
+		case "STRING_LITERAL":
+		case "STRING":
+			return typeNames.includes("string");
+		case "GENERIC_ARGUMENTS":
+			return matchesAtLeastOneNominalType(typeNames, typeB.target);
+		default:
+			return false;
+	}
+}
+
+function looksLikeADirective(type: SimpleType): boolean {
+	// If it has this type name, probably a directive.
+	if (type.name === "DirectiveFn") {
+		return true;
+	}
+	// Otherwise, look for a function that takes a part and returns nothing.
+	if (type.kind !== "FUNCTION" && type.kind !== "METHOD") {
+		return false;
+	}
+	if (type.argTypes === undefined || type.argTypes.length !== 1) {
+		return false;
+	}
+	const [argType] = type.argTypes;
+	if (argType.type.name !== "Part" && argType.type.name !== "NodePart") {
+		return false;
+	}
+	if (type.returnType === undefined || type.returnType.kind !== SimpleTypeKind.VOID) {
+		return false;
+	}
+	return true;
+}
```

Comment:
Is there a reason you didn't use `isLitDirective` from https://github.com/runem/lit-analyzer/blob/master/packages/lit-analyzer/src/analyze/util/directive/is-lit-directive.ts? Feel free to extend/refactor that function if needed :-)

Reply:
@rictic, any updates here? If you don't have time to combine the two functions, I'll merge this PR and combine them myself :blush: :+1:

Reply:
Yeah, just had to take a bit to get back to this. isLitDirective is more complete and careful, +1 to reusing it!